### PR TITLE
storage: Retry retry-able rocksdb errors

### PR DESCRIPTION
**Pull request by moulimukherjee**
Making changes so that we retry the retry-able errors in rocksdb. The maximum duration of the retries is configurable via a flag.

Tested the changes manually via the following scenarios:
- Changed code to throw RetryableError in the first iteration, as expected, it ran fine upon retry
- Changed code to always throw FatalError, and as expected it resulted into an error which eventually stalled the source
- Changed code to always throw RetryableError, and as expected it went above the max_duration and resulted into an error as well
- Also, checked that the `upsert_rocksdb_retry_duration` is truly configurable and could update it via `system-parameter-default`

### Motivation
Part of https://github.com/MaterializeInc/materialize/issues/17067

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/rocksdb/src/lib.rs [hunk 1]
```diff
@@ -524,26 +536,39 @@ fn rocksdb_core_loop<K, V, M, O>(
...
-                let gets: Result<Vec<_>, _> = gets.into_iter().collect();
-                match gets {
-                    Ok(gets) => {
-                        metrics.multi_get_latency.observe(latency.as_secs_f64());
-                        metrics.multi_get_size.observe(f64::cast_lossy(batch_size));
-                        let result = MultiGetResult {
-                            processed_gets: gets.len().try_into().unwrap(),
-                        };
+                let retry_result = Retry::default().max_tries(3).retry(|_| {
+                    let mut cloned_batch = batch.clone();
```
Comment (moulimukherjee) [inline src/rocksdb/src/lib.rs:548]: a clone here

File: src/rocksdb/src/lib.rs [hunk 2]
```diff
@@ -600,25 +622,41 @@ fn rocksdb_core_loop<K, V, M, O>(
                                     return;
                                 }
                             };
-                            writes.put(&key, encode_buf.as_slice());
+                            encoded_batch.push((key, Some(encode_buf.clone())));
```
Comment (moulimukherjee) [inline src/rocksdb/src/lib.rs:625]: another clone for each encoded_buf

Comment (guswynn) [reply src/rocksdb/src/lib.rs:625]: pseudo-code from talking online
```
let mut encoded_batch_buffers: Vec<Vec<u8>> = Vec::new();
let mut encoded_batch: Vec<(Key, Option<&[u8]>> = Vec::new();

while let Some(cmd) = cmd_rx.blocking_recv() {
   ...
}

for (key, value, encoded_buffer) in batch.drain(..).zip(encoded_batch_buffers.iter_mut() {
     // encode in `encoded_buffer`
     encoded_batch.push((key, Some(&encoded_buffer))
}

// retry logic around writing


for (i, _, encoded_buffer) in encoded_batch.drain(..).enumerate() {
    encoded_batch_buffers.insert(i, Some(encoded_buffer())
}
```

**Review by guswynn — APPROVED** [review]
one issue, but looks good otherwise!

File: src/rocksdb/src/lib.rs [hunk 3]
```diff
@@ -561,35 +590,42 @@ fn rocksdb_core_loop<K, V, M, O>(
...
 
                 let mut ret = MultiPutResult {
                     processed_puts: 0,
                     size_written: 0,
                 };
+
+                // initialize and push values into the buffer to match the batch size
+                let buf_size = encoded_batch_buffers.len();
+                for _ in buf_size..batch_size {
+                    encoded_batch_buffers.push(Some(Vec::new()));
```
[the remark this answers is not in the document]
Comment (moulimukherjee) [reply src/rocksdb/src/lib.rs:615]: Yeah and I added the assert below so that the length is at least of `batch_size`.

**Pull request merged**